wdl-lint/src/rules/shellcheck.rs

```diff
@@ -0,0 +1,387 @@
+    let mut decls = HashSet::new();
+    if let Some(cmd_parts) = section.strip_whitespace() {
+        cmd_parts.iter().for_each(|part| match part {
+            StrippedCommandPart::Text(text) => {
+                sanitized_command.push_str(text);
+            }
+            StrippedCommandPart::Placeholder(placeholder) => {
+                let bash_var = to_bash_var(placeholder);
+                // we need to save the var so we can suppress later
+                decls.insert(bash_var.clone());
+                let mut expansion = String::from("\"$");
+                expansion.push_str(&bash_var);
+                expansion.push('"');
+                sanitized_command.push_str(&expansion);
+            }
+        });
+        Some((sanitized_command, decls))
+    } else {
+        None
+    }
+}
+
+/// Returns the amount of leading whitespace characters in a `CommandSection`.
+///
+/// Only checks the first `CommandPart::Text`.
+fn count_command_whitespace(section: &CommandSection) -> usize {
+    if let Some(first_text) = section.parts().find(|p| matches!(p, CommandPart::Text(..))) {
+        match first_text {
+            CommandPart::Text(text) => {
+                let text_str = text
+                    .as_str()
+                    .strip_prefix("\n")
+                    .unwrap_or_else(|| text.as_str());
+                return count_leading_whitespace(text_str);
+            }
+            CommandPart::Placeholder(_) => unreachable!(),
+        }
+    }
+    0
+}
+
+impl Visitor for ShellCheckRule {
+    type State = Diagnostics;
+
+    fn document(
+        &mut self,
+        _: &mut Self::State,
+        reason: VisitReason,
+        _: &Document,
+        _: SupportedVersion,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        // Reset the visitor upon document entry
+        *self = Default::default();
+    }
+
+    fn command_section(
+        &mut self,
+        state: &mut Self::State,
+        reason: VisitReason,
+        section: &CommandSection,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        if !SHELLCHECK_EXISTS.get_or_init(|| {
+            if !program_exists(SHELLCHECK_BIN) {
+                let command_keyword = support::token(section.syntax(), SyntaxKind::CommandKeyword)
+                    .expect("should have a command keyword token");
+                state.exceptable_add(
+                    Diagnostic::error("running `shellcheck` on command section")
+                        .with_label(
+                            "could not find `shellcheck` executable.",
+                            command_keyword.text_range().to_span(),
+                        )
+                        .with_rule(ID)
+                        .with_fix("install shellcheck or disable this lint."),
+                    SyntaxElement::from(section.syntax().clone()),
+                    &self.exceptable_nodes(),
+                );
```

Comment:
Making a comment to open the floor for discussion.

Is this how we want to handle the case that SC isn't installed?

If we want to create a diagnostic that `ShellCheck` couldn't be run, it should be a `note` I think. _Not_ an `error`. Errors should be reserved for broken WDL, and this isn't that. I could maybe see this at `warning` level? But I'd prefer `note` personally.

Putting myself in the shoes of a naive user, if I've never heard of ShellCheck I'm just disabling this lint going forward. After all, we are asking folks to install another dependency without offering them any guidance on how to do that. And it would potentially annoy me seeing this every time I forget to disable the shellcheck rule. I think we need some more here. A link to shellcheck.net? And a one sentence sales pitch on the benefit of the SC?

Reply:
I thought we decided that ShellCheck was only going to be an opt-in rule? I think the behavior depends largely on that decision. I like making it opt-in because it delineates the behavior between silent failure because of the missing tool and just being turned off.

Reply:
> I thought we decided that ShellCheck was only going to be an opt-in rule? I think the behavior depends largely on that decision. I like making it opt-in because it delineates the behavior between silent failure because of the missing tool and just being turned off.

Maybe I missed something then. But the complication with making it opt-in is that this would be the first rule like that, and would probably require some tinkering of the architecture. We don't have a config system yet, so how exactly do we "opt-in"? Is that OOS for this PR?

Reply:
Well, I was thinking that this wouldn't be in the default set, and we would just add something in Sprocket that injects this rule in on a flag or similar. But, that being said, I'm fine with keeping it in the default set with no diagnostic on missing, so long as the TODO is marked in an issue (and, I would ask that, if we're close to having that working, we include it in the commit just commented out for now so it's easy to enable when we're ready).

Reply:
I'll go ahead and leave it in the default set but remove the diagnostic message on missing. I'll just comment out the diagnostic as requested (will fix up the wording a bit based on other feedback).